onnxruntime/contrib_ops/cpu/bert/decoder_masked_multihead_attention.cc

```diff
@@ -0,0 +1,470 @@
+      std::memcpy(present_value_data, past_value_data, past_value->SizeInBytes());
+    }
+
+    parameters.is_cross_attention = false;
+  }
+
+  if (output_qk_) {
+    int64_t qk_dims[] = {parameters.batch_size, parameters.num_heads, 1, parameters.total_sequence_length};
+    TensorShape qk_shape(&qk_dims[0], sizeof(qk_dims) / sizeof(qk_dims[0]));
+    cross_qk = context->Output(kQKOutputIndex, qk_shape);
+    parameters.out_qk = cross_qk->MutableData<T>();
+  }
+
+  // Beam width (in case we are using this op inside BeamSearch)
+  int beam_width_value = 1;
+  if (beam_width != nullptr) {
+    beam_width_value = static_cast<int>(*beam_width->Data<int32_t>());
+  }
+
+  // Cache indirection (in case we are using this op inside BeamSearch)
+  if (beam_width_value > 1 && cache_indir == nullptr) {
+    // If beam width > 1, then cache indirection buffer MUST be present
+    return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
+                           "If beam width is greater than 1, then cache indirection buffer MUST be present");
+  }
+
+  AllocatorPtr allocator;
+  ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&allocator));
+
+  OrtValue Q;
+  ORT_RETURN_IF_ERROR(MaybeTransposeToBNSHAndAddBias<T>(
+      context, allocator, batch_size, num_heads_, 1, head_size, query, bias, 0, Q));
+
+  // Cross-attention case
+  if (parameters.is_cross_attention) {
+    return ApplyAttention(Q.GetMutable<Tensor>()->MutableData<T>(),
+                          key->Data<T>(),
+                          value->Data<T>(),
+                          mask_index, nullptr /* past */, past_key, past_value, output, present_key, present_value,
+                          batch_size, 1 /* sequence_length */, parameters.kv_sequence_length,
+                          head_size, v_head_size, v_hidden_size, attention_bias, context, cross_qk);
+  }
+
+  OrtValue K, V;
+  ORT_RETURN_IF_ERROR(MaybeTransposeToBNSHAndAddBias<T>(
+      context, allocator, batch_size, num_heads_, 1, head_size, key, bias, hidden_size, K));
+  ORT_RETURN_IF_ERROR(MaybeTransposeToBNSHAndAddBias<T>(
+      context, allocator, batch_size, num_heads_, 1, v_head_size, value, bias, 2 * hidden_size, V));
+
+  // Self-attention, !has_beams
+  if (cache_indir == nullptr) {
+    return ApplyAttention(Q.GetMutable<Tensor>()->MutableData<T>(),
+                          K.GetMutable<Tensor>()->MutableData<T>(),
+                          V.GetMutable<Tensor>()->MutableData<T>(),
+                          mask_index, nullptr /* past */, past_key, past_value, output, present_key, present_value,
+                          batch_size, 1 /* sequence_length */, parameters.kv_sequence_length,
+                          head_size, v_head_size, v_hidden_size, attention_bias, context, cross_qk,
+                          parameters.past_sequence_length, true /* past_present_share_buffer */);
+  }
+
+  // Self-attention, has_beams
+  return ApplyAttentionWithBeams(Q.GetMutable<Tensor>()->MutableData<T>(),
```

Comment:
Can this method be moved to a header file such as `attention_cpu_base.h`? This method will eventually need to be called by the MHA op internally.